exercises/src/main/scala/fpinscala/datastructures/List.scala

```diff
@@ -50,19 +52,112 @@ object List { // `List` companion object. Contains functions for creating and wo
+    case Nil => Nil
+    case Cons(h, t) =>
+      if (f(h)) dropWhile(t, f)
+      else l
+  }
+
+  def init[A](l: List[A]): List[A] = l match {
+      case Nil => Nil
+      case Cons(_, Nil) => Nil
+      case Cons(h, t) => Cons(h, init(t))
+  }
+
+  def length[A](l: List[A]): Int = l match {
+    case Nil => 0
+    case Cons(_, Nil) => 1
+    case Cons(_, t) => 1 + length(t)
+  }
+
+  //length with foldRight
+  def length2[A](l: List[A]): Int = foldRight(l, 0)((_: A, b: Int) => b + 1)
+
+  @tailrec
+  def foldLeft[A,B](l: List[A], z: B)(f: (A, B) => B): B = l match {
+    case Nil => z
+    case Cons(h, t) => foldLeft(t, f(h, z))(f)
+  }
 
-  def setHead[A](l: List[A], h: A): List[A] = ???
+  def sumFL(ns: List[Int]) =
+    foldLeft(ns, 0)((x,y) => x + y)
 
-  def drop[A](l: List[A], n: Int): List[A] = ???
+  def productFL(ns: List[Double]) =
+    foldLeft(ns, 1.0)(_ * _)
 
-  def dropWhile[A](l: List[A], f: A => Boolean): List[A] = ???
+  def reverse[A](l: List[A]): List[A] = foldLeft(l, List[A]())((y, x) => Cons(y, x))
 
-  def init[A](l: List[A]): List[A] = ???
+  def foldRightTailRec[A,B](l: List[A], z: B)(f: (A, B) => B): B = foldLeft(reverse(l), z)(f)
 
-  def length[A](l: List[A]): Int = ???
+  //append using fold
+  def appendFold[A](a1: List[A], a2: List[A]): List[A] = foldRight(a1, a2)((x, y) => Cons[A](x, y))
 
-  def foldLeft[A,B](l: List[A], z: B)(f: (B, A) => B): B = ???
+  //concat a list of lists
+  def concatenate[A](ls: List[List[A]]): List[A] = foldLeft(ls, List[A]())((x, y) => appendFold(y, x))
 
-  def map[A,B](l: List[A])(f: A => B): List[B] = ???
+  //add 1 to a list of ints
+  def addOne(l: List[Int]): List[Int] = foldRightTailRec(l, List[Int]())((x, y) => Cons(x + 1, y))
+
+  def doublesToStrings(l: List[Double]): List[String] = foldRightTailRec(l, List[String]())((x, y) => append(List(x.toString), y))
+
+  def map[A,B](l: List[A])(f: A => B): List[B] = foldRightTailRec(l, List[B]())((a, b) => append(List(f(a)), b))
+
+  def filter[A](l: List[A])(f: A => Boolean): List[A] = foldRightTailRec(l, List[A]()) {
+    (a, b) => if (f(a)) append(List(a), b) else b
+  }
+
+  def flatMap[A, B](l: List[A])(f: A => List[B]): List[B] = foldRightTailRec(l, List[B]())((a, b) => append(f(a), b))
+
+  def filterFlatMap[A](l: List[A])(f: A => Boolean): List[A] = flatMap(l)(a => if (f(a)) List(a) else Nil)
+
+  //add the elements of 2 lists
+  def addLists(l1: List[Int], l2: List[Int]): List[Int] = (l1, l2) match {
+    case (_, Nil) => Nil
+    case (Nil, _) => Nil
+    case (Cons(h1, t1), Cons(h2, t2)) => Cons(h1 + h2, addLists(t1, t2))
+  }
+
+  //generalize addLists
+  def zipWith[A, B, C](l1: List[A], l2: List[B])(f: (A, B) => C): List[C] = (l1, l2) match {
+    case (_, Nil) => Nil
+    case (Nil, _) => Nil
+    case (Cons(h1, t1), Cons(h2, t2)) => Cons(f(h1, h2), zipWith(t1, t2)(f))
+  }
+
+  def hasSubsequence[A](l: List[A], sub: List[A]): Boolean = {
+    def startsWith(sub: List[A], l: List[A]): Boolean = sub match {
```

Comment:
nit: It's better to avoid shadowing variable names like this. If you rename `sub`, then the code would still compile but behave differently. It's preferable to keep names unique. Most of our project will have compiler flags added so that shadowing like this wouldn't compile